pkg/virtual/framework/rootapiserver/root_apiserver.go

```diff
@@ -0,0 +1,224 @@
+		if vwNames.Has(name) {
+			return nil, errors.New("Several virtual workspaces with the same name: " + name)
+		}
+		vwNames.Insert(name)
+
+		var err error
+		delegateAPIServer, err = virtualWorkspace.Register(c.GenericConfig, delegateAPIServer)
+		if err != nil {
+			return nil, err
+		}
+		readys = append(readys, virtualWorkspace.IsReady)
+	}
+
+	c.GenericConfig.BuildHandlerChainFunc = c.getRootHandlerChain(delegateAPIServer)
+	c.GenericConfig.RequestInfoResolver = c
+	c.GenericConfig.AddHealthChecks(Readys(readys))
+
+	genericServer, err := c.GenericConfig.New("virtual-workspaces-root-apiserver", delegateAPIServer)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &RootAPIServer{
+		GenericAPIServer: genericServer,
+	}
+
+	// register our poststarthooks
+	s.GenericAPIServer.AddPostStartHookOrDie("virtual-workspace-startinformers", func(context genericapiserver.PostStartHookContext) error {
+		c.ExtraConfig.informerStart(context.StopCh)
+		return nil
+	})
+
+	return s, nil
+}
+
+type Readys []framework.ReadyFunc
+
+func (r Readys) Name() string {
+	return "VirtualWotrkspaceAdditionalReadinessChecks"
+}
+
+func (readys Readys) Check(req *http.Request) error {
+	for _, ready := range readys {
+		if err := ready(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c completedConfig) resolveRootPaths(urlPath string, requestContext context.Context) (accepted bool, prefixToStrip string, completedContext context.Context) {
+	completedContext = requestContext
+	for _, virtualWorkspace := range c.ExtraConfig.VirtualWorkspaces {
+		if accepted, prefixToStrip, completedContext := virtualWorkspace.ResolveRootPath(urlPath, requestContext); accepted {
+			return accepted, prefixToStrip, context.WithValue(completedContext, virtualcontext.VirtualWorkspaceNameKey, virtualWorkspace.GetName())
+		}
+	}
+	return
+}
+
+func (c completedConfig) getRootHandlerChain(delegateAPIServer genericapiserver.DelegationTarget) func(http.Handler, *genericapiserver.Config) http.Handler {
+	return func(apiHandler http.Handler, genericConfig *genericapiserver.Config) http.Handler {
+		return genericapiserver.DefaultBuildHandlerChain(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			if accepted, prefixToStrip, context := c.resolveRootPaths(req.URL.Path, req.Context()); accepted {
+				req.URL.Path = strings.TrimPrefix(req.URL.Path, prefixToStrip)
+				req.URL.RawPath = strings.TrimPrefix(req.URL.RawPath, prefixToStrip)
+				req = req.WithContext(genericapirequest.WithCluster(context, genericapirequest.Cluster{Name: "virtual"}))
+				delegatedHandler := delegateAPIServer.UnprotectedHandler()
+				if delegatedHandler != nil {
+					delegatedHandler.ServeHTTP(w, req)
+				}
+				return
+			}
+			apiHandler.ServeHTTP(w, req)
+		}), c.GenericConfig.Config)
+	}
+}
+
+var _ genericapirequest.RequestInfoResolver = (*completedConfig)(nil)
+
+func (c completedConfig) NewRequestInfo(req *http.Request) (*genericapirequest.RequestInfo, error) {
```

Comment:
why is this called?

Reply:
It is expected to be called by the `RootAPIServer` in order to  remove from the request the prefix associated to an accepted virtual workspace root path, so that the `RequestInfo` is not created from an invalid URL path.
Since we reuse the `DefaultBuildChainHandler` in our customized `getRootChainHandler` (that checks the URL Path against virtual workspace root paths), that means that the `RequestInfo` will be created first (as part of the `DefaultBuildChainHandler` ). So we also override the `RequestInfoResolver` so that it would use the same URL Path as the one that will be forwarded to the virtual workspace deletegated APIServers.

Reply:
didn't find a reference. Maybe Chrome hadn't rendered the whole page yet. All fine.

Reply:
With this method the `completedConfig` is made a `RequestInfoResolver`
And some lines above with have:
```go
	c.GenericConfig.RequestInfoResolver = c
```
So this method will finally be called by the `RootAPIServer`.

I added a comment to describe this in more details.